gateway/src/apicast/policy/metrics/metrics.lua

```diff
@@ -0,0 +1,113 @@
+local _M = require('apicast.policy').new('Metrics')
+
+local resty_env = require('resty.env')
+local errlog = require('ngx.errlog')
+local prometheus = require('apicast.prometheus')
+local metrics_updater = require('apicast.metrics.updater')
+local tonumber = tonumber
+local select = select
+local find = string.find
+local pairs = pairs
+
+local new = _M.new
+
+local log_map = {
+  'emerg',
+  'alert',
+  'crit',
+  'error',
+  'warn',
+  'notice',
+  'info',
+  'debug',
+}
+
+local log_level_env = 'METRICS_LOG_LEVEL'
+local max_logs_env = 'METRICS_MAX_LOGS'
+
+local log_level_default = 'error'
+local max_logs_default = 100
+
+local function find_i(t, value)
+  for i=1, #t do
+    if t[i] == value then return i end
+  end
+end
+
+local empty = {}
+
+local function get_logs(max)
+  return errlog.get_logs(max) or empty
+end
+
+local function filter_level()
+  local level = resty_env.value(log_level_env) or log_level_default
+
+  local level_index = find_i(log_map, level)
+
+  if not level_index then
+    ngx.log(ngx.WARN, _M._NAME, ': invalid level: ', level, ' using error instead')
+    level_index = find_i(log_map, 'error')
+  end
+
+  return level_index
+end
+
+function _M.new(configuration)
+  local m = new()
+
+  local config = configuration or empty
+
+  -- how many logs to take in one iteration
+  m.max_logs = tonumber(config.max_logs) or
+               resty_env.value(max_logs_env) or
+               max_logs_default
+
+  return m
+end
+
+local logs_metric = prometheus('counter', 'nginx_error_log', "Items in nginx error log", {'level'})
+local http_connections_metric =  prometheus('gauge', 'nginx_http_connections', 'Number of HTTP connections', {'state'})
+local shdict_capacity_metric = prometheus('gauge', 'openresty_shdict_capacity', 'OpenResty shared dictionary capacity', {'dict'})
+local shdict_free_space_metric = prometheus('gauge', 'openresty_shdict_free_space', 'OpenResty shared dictionary free space', {'dict'})
+
+function _M.init()
+  errlog.set_filter_level(filter_level())
+
+  get_logs(100) -- to throw them away after setting the filter level (and get rid of debug ones)
+
+  for name,dict in pairs(ngx.shared) do
+    metrics_updater.set(shdict_capacity_metric, dict:capacity(), name)
+  end
+end
+
+function _M:metrics()
+  local logs = get_logs(self.max_logs)
+
+  for i = 1, #logs, 3 do
```

Comment:
Each of those 3 different metrics could be own function to reduce complexity.

Reply:
I'd rather address refactorings in a future PR.

Reply:
👍